{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Read in user provided file\n",
    "Support: ionbot (ionbot.modifications.csv), MSFragger (psm.tsv), open pFind (.proteins)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [],
   "source": [
    "import re\n",
    "import pandas as pd\n",
    "from Bio import SeqIO\n",
    "from unimod_mapper import UnimodMapper"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [],
   "source": [
    "user_flag = \"ionbot\" # here goes the thing the user selected from the dropdown. has to be one of [ionbot, msfragger, openpfind]\n",
    "user_file = \"example_input/ionbot.modifications.csv\" # here goes the path to the users file\n",
    "# fasta = \"fastas/uniprot_sprot.fasta\" # this is the fasta with the protein sequences. if the user doesnt provide one, we assume the protein IDs are in swissprot and use the uniprot_sprot.fasta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [],
   "source": [
    "mapper = UnimodMapper()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [],
   "source": [
    "uniprot_to_sequence = {}\n",
    "for record in SeqIO.parse(fasta, \"fasta\"):\n",
    "    sequence = record.seq\n",
    "    uniprot_id = record.id.split(\"|\")[1]\n",
    "    uniprot_to_sequence[uniprot_id] = str(sequence)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_ptm_position_in_protein(mod, protein_start):\n",
    "    position_ptm_in_peptide = int(re.search(\".+?(?=\\D)\", mod)[0]) \n",
    "    position_ptm_in_protein = protein_start + position_ptm_in_peptide - 1\n",
    "    return position_ptm_in_protein"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_file(file, flag):\n",
    "    if flag == \"ionbot\":\n",
    "        df = read_ionbot(file)\n",
    "    elif flag == \"msfragger\":\n",
    "        df = read_msfragger(file)\n",
    "    elif flag == \"openpfind\":\n",
    "        df = read_openpfind(file)\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_ionbot(file):\n",
    "    df = pd.read_csv(file)\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [],
   "source": [
    "def map_mass_to_unimod(mods):\n",
    "    modstr = \"\"\n",
    "    for mod in mods.split(\",\"):\n",
    "        pos_aa = mod.split(\"(\")[0]\n",
    "\n",
    "        mass = re.search(r'\\(.*?\\)', mod)[0]\n",
    "        mass = float(mass.replace(\"(\", \"\").replace(\")\", \"\"))\n",
    "        mapped = mapper.mass_to_names(mass, decimals=4) # for unimod ID: mapper.mass_to_ids(mass, decimals = 4)\n",
    "\n",
    "        mapped_names = \" or \".join(list(mapped))\n",
    "        if len(list(mapped)) > 0:\n",
    "            modstr += pos_aa + \"[\" + mapped_names + \"]\"\n",
    "\n",
    "    return modstr\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {},
   "outputs": [],
   "source": [
    "def from_psm_to_protein(df):\n",
    "    # for each psm in msfragger, get protein sequence via uniprot ID and map PTM position onto it\n",
    "    uniprot_ids = []\n",
    "    modifications = []\n",
    "    positions = []\n",
    "    \n",
    "    for index, psm in df.iterrows():\n",
    "        protein_start = psm[\"Protein Start\"]\n",
    "        uniprot_id = psm[\"Protein ID\"]\n",
    "        \n",
    "        for mod in psm[\"Assigned Modifications parsed\"].split(\"]\"):\n",
    "            if mod == \"\":\n",
    "                continue\n",
    "            modification = mod.split(\"[\")[1]\n",
    "            position = get_ptm_position_in_protein(mod, protein_start)\n",
    "\n",
    "            modifications.append(modification)\n",
    "            positions.append(position)\n",
    "            uniprot_ids.append(uniprot_id)\n",
    "            \n",
    "    zipped = list(zip(uniprot_ids, modifications, positions))\n",
    "    protein_df = pd.DataFrame(zipped, columns=[\"uniprot_id\",\"modification\",\"position\"])\n",
    "\n",
    "    return protein_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_msfragger(file):\n",
    "    # read file and pick relevant rows and columns\n",
    "    print(\"Reading MSFragger file...\")\n",
    "    pept_mods = pd.read_csv(file, sep = \"\\t\")\n",
    "    pept_mods = pept_mods[[\"Modified Peptide\", \"Peptide\", \"Protein\", \"Protein ID\", \"Assigned Modifications\", \"Protein Start\"]]\n",
    "    pept_mods = pept_mods.dropna(subset=[\"Modified Peptide\"])\n",
    "    pept_mods = pept_mods.drop_duplicates()\n",
    "\n",
    "    # try to assign modifications to mass shifts\n",
    "    print(\"Assigning modifications to mass shifts...\")\n",
    "    pept_mods[\"Assigned Modifications parsed\"] = pept_mods[\"Assigned Modifications\"].apply(lambda x: map_mass_to_unimod(x)) \n",
    "    \n",
    "    # rewrite to protein level modifications\n",
    "    print(\"Mapping modification sites onto proteins...\")\n",
    "    prot_mods = from_psm_to_protein(pept_mods)\n",
    "\n",
    "    return prot_mods"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_openpfind(file):\n",
    "\n",
    "    with open(file, 'w') as openfile:\n",
    "        lines = openfile.readlines()\n",
    "        print(lines)\n",
    "        \n",
    "    \n",
    "    rows = file.read_lines()\n",
    "\n",
    "    # Initialize a dictionary to hold the parsed data\n",
    "    data = {}\n",
    "\n",
    "    # Loop over the rows\n",
    "    for i, row in enumerate(rows):\n",
    "        print(row)\n",
    "        \n",
    "        # Split the row into cells\n",
    "        cells = row.split('\\t')\n",
    "        \n",
    "        # If this is the first row, set the column names\n",
    "        if i == 0:\n",
    "            columns = cells\n",
    "        \n",
    "        # Otherwise, add the row to the data dictionary\n",
    "        else:\n",
    "            row_data = {}\n",
    "            for j, cell in enumerate(cells):\n",
    "                row_data[columns[j]] = cell\n",
    "            data[i] = row_data\n",
    "\n",
    "\n",
    "\n",
    "    print(psm_df)\n",
    "    return"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {},
   "outputs": [],
   "source": [
    "# try to parse user file\n",
    "try:\n",
    "    df = read_file(user_file, user_flag)\n",
    "except TabError:\n",
    "    print(\"Your file could not be parsed. Have you selected the right format?\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protein</th>\n",
       "      <th>uniprot_id</th>\n",
       "      <th>unexpected_modification</th>\n",
       "      <th>position</th>\n",
       "      <th>ionbot_match_id</th>\n",
       "      <th>#PSMs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3PASE_ECOLI</td>\n",
       "      <td>P30871</td>\n",
       "      <td>[7]Deamidated[N]</td>\n",
       "      <td>61</td>\n",
       "      <td>0_13437_1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>3PASE_ECOLI</td>\n",
       "      <td>P30871</td>\n",
       "      <td>[10]Met-&gt;Hse[M]</td>\n",
       "      <td>266</td>\n",
       "      <td>0_53627_2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>5DNU_ECOLI</td>\n",
       "      <td>P76491</td>\n",
       "      <td>[35]Oxidation[E]</td>\n",
       "      <td>168</td>\n",
       "      <td>0_14780_2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>6PGD_ECOLI</td>\n",
       "      <td>P00350</td>\n",
       "      <td>[35]oxidation[M]</td>\n",
       "      <td>11</td>\n",
       "      <td>0_35178_2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>6PGD_ECOLI</td>\n",
       "      <td>P00350</td>\n",
       "      <td>[9999575]SulfanilicAcid[E]</td>\n",
       "      <td>167</td>\n",
       "      <td>0_44305_1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10060</th>\n",
       "      <td>sp|TRYP_PIG|</td>\n",
       "      <td>sp|TRYP_PIG|</td>\n",
       "      <td>[34]Methyl[N-TERM]</td>\n",
       "      <td>108</td>\n",
       "      <td>0_23420_2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10061</th>\n",
       "      <td>sp|TRYP_PIG|</td>\n",
       "      <td>sp|TRYP_PIG|</td>\n",
       "      <td>[255]Delta:H(4)C(2)[N-TERM]</td>\n",
       "      <td>108</td>\n",
       "      <td>0_24698_2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10062</th>\n",
       "      <td>sp|TRYP_PIG|</td>\n",
       "      <td>sp|TRYP_PIG|</td>\n",
       "      <td>[254]Delta:H(2)C(2)[N-TERM]</td>\n",
       "      <td>108</td>\n",
       "      <td>0_33029_1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10063</th>\n",
       "      <td>sp|TRYP_PIG|</td>\n",
       "      <td>sp|TRYP_PIG|</td>\n",
       "      <td>[385]Ammonia-loss[T]</td>\n",
       "      <td>110</td>\n",
       "      <td>0_22152_1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10064</th>\n",
       "      <td>sp|TRYP_PIG|</td>\n",
       "      <td>sp|TRYP_PIG|</td>\n",
       "      <td>[23]Dehydrated[T]</td>\n",
       "      <td>110</td>\n",
       "      <td>0_22622_1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10065 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            protein    uniprot_id      unexpected_modification  position  \\\n",
       "0       3PASE_ECOLI        P30871             [7]Deamidated[N]        61   \n",
       "1       3PASE_ECOLI        P30871              [10]Met->Hse[M]       266   \n",
       "2        5DNU_ECOLI        P76491             [35]Oxidation[E]       168   \n",
       "3        6PGD_ECOLI        P00350             [35]oxidation[M]        11   \n",
       "4        6PGD_ECOLI        P00350   [9999575]SulfanilicAcid[E]       167   \n",
       "...             ...           ...                          ...       ...   \n",
       "10060  sp|TRYP_PIG|  sp|TRYP_PIG|           [34]Methyl[N-TERM]       108   \n",
       "10061  sp|TRYP_PIG|  sp|TRYP_PIG|  [255]Delta:H(4)C(2)[N-TERM]       108   \n",
       "10062  sp|TRYP_PIG|  sp|TRYP_PIG|  [254]Delta:H(2)C(2)[N-TERM]       108   \n",
       "10063  sp|TRYP_PIG|  sp|TRYP_PIG|         [385]Ammonia-loss[T]       110   \n",
       "10064  sp|TRYP_PIG|  sp|TRYP_PIG|            [23]Dehydrated[T]       110   \n",
       "\n",
       "      ionbot_match_id  #PSMs  \n",
       "0           0_13437_1      1  \n",
       "1           0_53627_2      1  \n",
       "2           0_14780_2      1  \n",
       "3           0_35178_2      1  \n",
       "4           0_44305_1      1  \n",
       "...               ...    ...  \n",
       "10060       0_23420_2      1  \n",
       "10061       0_24698_2      1  \n",
       "10062       0_33029_1      1  \n",
       "10063       0_22152_1      1  \n",
       "10064       0_22622_1      1  \n",
       "\n",
       "[10065 rows x 6 columns]"
      ]
     },
     "execution_count": 87,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "proteomics",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.10"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
