<!DOCTYPE html>
<!--
PTMVision WEB APPLICATION
AUTHOR: Caroline Jachmann, Simon Hackl
CONTACT: caroline.jachmann@uni-tuebingen.de, simon.hackl@uni-tuebingen.de
VERSION: 1.0

LICENSE
This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.
This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.
You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.
-->
<html>
  <head>
    <title>About - PTMVision</title>
    <link
      rel="icon"
      type="image/x-icon"
      href="{{ url_for('static', filename='resources/icon.png') }}"
    />
    <!--
      Provide meta information about the application.
      -->
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <meta
      name="description"
      content="PTMVision: Interactive Visualization of Post Translational Modifications | Understanding the intricate landscape of PTMs is crucial for unraveling the complex mechanisms underlying cellular processes and diseases. With PTMVision, researchers can effortlessly analyze and explore PTM data derived from various sources, including search engines such as ionbot, MSFragger, Sage, MaxQuant, or from a plain CSV table format processed from mass spectrometry (MS) data."
    />
    <meta name="author" content="Caroline Jachmann, Simon Hackl" />
    <!--
      Definition of stylesheets to use in the application.
      -->
    <link
      rel="stylesheet"
      href="https://cdn.korzh.com/metroui/v4/css/metro-all.min.css"
    />
    <link
      rel="stylesheet"
      type="text/css"
      href="{{url_for('static', filename='css/index.css')}}"
    />
    <!--
      Definition of java script files to use in the application.
      -->
    <script
      type="text/javascript"
      src="https://code.jquery.com/jquery-3.6.0.min.js"
    ></script>
    <script
      type="text/javascript"
      src="https://cdn.korzh.com/metroui/v4/js/metro.min.js"
    ></script>
    <script
      type="text/javascript"
      src="https://kit.fontawesome.com/df04f22e3d.js"
      crossorigin="anonymous"
    ></script>
    <script
      type="text/javascript"
      src="{{url_for('static', filename='js/index.js')}}"
    ></script>
  </head>
  <body onload='$("#menu")[0].style.display = "flex";'>
    <!-- App menu -->
    <header
      id="menu"
      class="app-bar app-bar-expand"
      data-role="appbar"
      data-expand="true"
      data-role-appbar="true"
    >
      <div id="menu-grid" class="grid w-100">
        <div class="row">
          <div class="cell text-left ml-4 w-100">
            <div class="logo no-hover">
              <img
                src="{{ url_for('static', filename='resources/logo.png') }}"
                alt="PTMVision Logo"
              />
            </div>
          </div>
        </div>
      </div>
    </header>

    <div id="main">
      <div
        class="p-8"
        data-role="accordion"
        data-one-frame="true"
        data-show-active="true"
      >
        <div class="frame active">
          <div class="heading">About PTMVision</div>
          <div class="content">
            <h4>
              <small>
                PTMVision (Interactive Visualization of Post Translational
                Modifications) is an interactive web platform for effortless
                exploratory visual analysis of
                <a
                  href="https://www.sciencedirect.com/topics/biochemistry-genetics-and-molecular-biology/posttranslational-modification"
                  >post translational modifications (PTMs)</a
                >
                obtained from
                <a
                  href="https://www.sciencedirect.com/topics/biochemistry-genetics-and-molecular-biology/mass-spectrometry"
                  >mass spectrometry (MS)</a
                >
                data using search engines such as
                <a href="https://ionbot.cloud/">ionbot</a>,
                <a href="https://www.maxquant.org/">MaxQuant</a>,
                <a
                  href="https://www.google.com/search?client=firefox-b-d&q=MsFragger"
                  >MSFragger</a
                >, or <a href="https://sage-docs.vercel.app/docs">Sage</a>. It
                enables researchers to comprehend the intricate landscape of
                PTMs in order to unravel possible mechanisms related to cellular
                processes and diseases.
              </small>
            </h4>
            <h4><i class="fa-solid fa-star-shooting"></i> Key Features</h4>
            <ul>
              <li>
                <code>Variety of overview and summary visualizations</code> that
                help identify patterns and trends in PTM data.
              </li>
              <li>
                Visualizations are dynamically linked and support features to
                highlight PTMs of interest.
              </li>
              <li>
                <code
                  >Integration of protein structure and annotation data from
                  UniProt</code
                >
                provide multidimensional view. This improves the
                interpretability of the data is significantly.
              </li>
              <li>
                <code
                  >Unique emphasis on interactions of residues in close contact
                  within a single protein</code
                >
                and associated modifications. This allows to study the dynamic
                interplay between PTMs and their putative effects on protein
                folding and function.
              </li>
              <li>
                <code>Responsive and intuitive interface</code> that allows
                users to effortlessly navigate the website and through their
                data.
              </li>
              <li>
                <code>Supports high degree of data provenance</code> by allowing
                various input formats and an easy export of session data to
                continue analysis without repeated processing input data.
              </li>
            </ul>
          </div>
        </div>

        <div class="frame">
          <div class="heading">Usage Guide</div>
          <div class="content">
            <h6>Content</h6>
            <ul>
              <li>Starting a PTMVision Session</li>
              <ul class="aboute-content-list">
                <li onclick="to('about-c1')">Uploading Data</li>
                <li onclick="to('about-c2')">Retrieving Session Data</li>
                <li onclick="to('about-c3')">Supported Data Formats</li>
              </ul>
              <li>Obtaining an Overview of Analysis Results</li>
              <ul class="aboute-content-list">
                <li onclick="to('about-c4')">Overview Visualizations</li>
                <li onclick="to('about-c5')">Shared PTM sites between modification types</li>
                <li onclick="to('about-c6')">Mass Shift</li>
                <li onclick="to('about-c7')">Site Count</li>
                <li onclick="to('about-c8')">PTM Class Counts</li>
                <li onclick="to('about-c9')">Overview Table</li>
              </ul>
              <li>Investigating the PTMs of one Protein in Detail</li>
              <ul class="aboute-content-list">
                <li onclick="to('about-c10')">Detail Visualizations</li>
                <li onclick="to('about-c11')">Modifications View</li>
                <li onclick="to('about-c12')">Structure View</li>
              </ul>
            </ul>
            <hr />
            <h6 id="about-c1"><strong>1 Uploading Data</strong></h6>
            <small
              >On the landing page at panel
              <i class="fa-duotone fa-circle-1 fa-sm"></i> users have three
              options to start a session with PTMVision.
              <br />
              <strong>(1)</strong> Start a novel analysis by entering data in
              the required inputs
              <em>Specify pipeline with which your data was generated</em>
              and <em>Specify PTM data file</em> and clicking on the
              <span
                style="
                  background-color: #ff6663;
                  padding: 0.5px;
                  margin: 1px;
                  font-size: small;
                  border-radius: 4px;
                  color: #fbfbfb;
                "
                >Submit</span
              >
              button.
              <br />
              <strong>(2)</strong> Re-upload the data of a previous analysis by
              filling in the form
              <em
                >Specify previously downloaded PTMVision session .zlib file</em
              >
              and clicking the corresponding
              <span
                style="
                  background-color: #ff6663;
                  padding: 0.5px;
                  margin: 1px;
                  font-size: small;
                  border-radius: 4px;
                  color: #fbfbfb;
                "
                >Submit</span
              >
              button. Please ensure, that the data provided is downloaded with
              PTMVision. A description of how you can save your session data is
              given in the next section.
              <br />
              <strong>(3)</strong> Load a small example dataset via the provided
              link <em>load an example session</em> in the text to dive directly
              into PTMVision.</small
            >
            <hr />
            <h6 id="about-c2"><strong>2 Retrieving Session Data</strong></h6>
            <small
              >Once you have processed your data with PTMVision you can download
              the results in a binary <em>zlib</em> compressed format by
              clicking on <i class="fa-duotone fa-file-arrow-down"></i> in the
              header menu on the right. The downloaded format is not human
              readable, but contains all relevant information of your session.
              The file can be re-uploaded to re-initialize a session and is also
              intended to easily share results with fellow users.</small
            >
            <hr />
            <h6 id="about-c3"><strong>3 Supported Data Formats</strong></h6>
            <small>
              We currently support the following input formats; If you can't
              find your favorite search engine, feel free to
              <a
                href="https://github.com/Integrative-Transcriptomics/PTMVision/issues/new"
                ,
                target="_blank"
                >open an issue</a
              >.
              <table class="table striped">
                <thead>
                  <tr>
                    <th>Source</th>
                    <th>Version Tested</th>
                    <th>Columns Used</th>
                    <th>Postprocessing</th>
                  </tr>
                </thead>
                <tbody>
                  <tr>
                    <td>ionbot</td>
                    <td><code>v0.10.0</code>, <code>v0.11.0</code></td>
                    <td>uniprot_id, unexpected_modification, position</td>
                    <td>N/A</td>
                  </tr>
                  <tr>
                    <td>Sage</td>
                    <td><code>v0.13.1</code></td>
                    <td>*</td>
                    <td>
                      1% FDR Filtering (PSM level); Remove decoys; Remove
                      peptides matching to more than one protein; Map mass
                      shifts to UniMod IDs; Map PTM position onto protein; Mass
                      shifts are not localised, only variable mods are used.
                    </td>
                  </tr>
                  <tr>
                    <td>FragPipe</td>
                    <td>
                      <code>MSFragger v4.0</code>,
                      <code>PTMShepherd v2.0.6</code>
                    </td>
                    <td>
                      Modified Peptide, Peptide, Protein ID, Assigned
                      Modifications, Observed Modifications, MSFragger
                      Localization, Protein Start
                    </td>
                    <td>
                      Filter ambiguous localisations; Filter ambiguous
                      modifications; Filter combinations of modifications; Map
                      mass shifts and UniMod Descriptions to UniMod IDs; Map PTM
                      position onto protein; Retrieve UniMod Classification; **.
                    </td>
                  </tr>
                  <tr>
                    <td>MaxQuant</td>
                    <td>
                      <code>v2.4.13.0</code>
                    </td>
                    <td>*</td>
                    <td>
                      Remove peptides matching to more than one protein; Map
                      MaxQuant modification names to UniMod names; Map PTM
                      position onto protein.
                    </td>
                  </tr>
                  <tr>
                    <td>Plain CSV</td>
                    <td>N/A</td>
                    <td>***</td>
                    <td>
                      If no classification given: Retrieve protein sequence,
                      extract modified amino acid, retrieve UniMod
                      classification.
                    </td>
                  </tr>
                </tbody>
              </table>
              <div>
                * See
                <a
                  href="https://psm-utils.readthedocs.io/en/v0.6.0/"
                  target="_blank"
                  >psm-utils.readthedocs.io/en/v0.6.0/</a
                >
                for details
                <br />
                ** For FragPipe, localisations for unexpected modifications are
                derived from the "MSFragger Localization" column. This
                localisation might be only marginally better than the next best
                site, so caution is advised. As soon as the next MSFragger
                version is released, which reports localisation scores as well,
                PTMVision will include a way to filter modifications based on
                the scores.
                <br />
                *** <em>uniprot_id</em>: UniProt Accession of the protein;
                <em>position</em>: Position of the modification in the protein
                sequence (1 indexed); <em>modification_unimod_name</em>: UniMod
                Name; <em>modification_unimod_id</em>: UniMod ID; (optional)
                <em>mass_shift</em>: mass of the modification in Dalton (will be
                inferred from unimod ID if not provided); (optional)
                <em>classification</em>: unimod classification (will be inferred
                from unimod ID and modified residue if not provided)
              </div>
            </small>
            <hr />
            <h6 id="about-c4"><strong>4 Overview Visualizations</strong></h6>
            <small>
              After successful data processing, you will be presented with a
              variety of overview visualizations (described in detail in
              sections five to eight) and a tabular overview of all proteins in
              your data (described in detail in section nine). All overview
              visualizations have zoom and pan operations, for which the y-axes
              of those visualizations that display information per modification
              are linked, and offer a range of user interactions accessible via
              the buttons at the top right of panel
              <i class="fa-duotone fa-circle-2"></i>:
              <p>
                <i class="fa-duotone fa-camera"></i> Screenshot of the current
                view of the overview visualizations. The corresponding image
                file is automatically downloaded in <code>png</code> format.
              </p>
              <p>
                <i class="fa-duotone fa-magnifying-glass-arrows-rotate"></i>
                Reset the zoom level of all overview visualizations to default
                (zoomed out completely).
              </p>
              <p>
                <i class="fa-duotone fa-highlighter"></i> Highlight a specific
                PTM of interest. A pop-up window that supports search input
                prompts you to select a number of modifications to be
                highlighted. When you confirm the popup, lines are drawn at the
                respective coordinates of the specified modifications. To remove
                a highlight, the interaction must be performed without selecting
                a modification.
              </p>
              <p>
                <i class="fa-duotone fa-arrow-up-arrow-down"></i> Sort the
                <em>Shared PTM sites</em>, <em>Mass Shift</em> and
                <em>Site Count</em> visualizations either by mass shift or
                count. By default, the axes are sorted by count. If sorting by
                mass shift is set, red areas in the
                <em>Mass Shift</em> visualization will be highlighted to
                indicate modifications whose mass shift difference is less than
                <code>0.02 Da</code>.
              </p>
              <!--<div class="img-container rounded thumbnail mt-4 mb-4">
                <img
                  src="{{ url_for('static', filename='resources/usage_guide_images/overview_1.png') }}"
                  alt="Default view of PTM overview visualizations, sorted by Per PTM Count, no highlight."
                />
                <span class="title"
                  ><secondary
                    >Default view of PTM overview visualizations, sorted by
                    <em>Per PTM Count</em>, no highlight.</secondary
                  ></span
                >
              </div>-->

              <div class="img-container thumbnail">
                <video controls="controls" style="width: 100%;" name="Video Name" poster="{{ url_for('static', filename='resources/usage_guide_images/overview_2.png') }}">
                  <source src="{{ url_for('static', filename='resources/usage_guide_movies/Overview_PTMs.mov') }}"">
                </video>
                <span class="title"
                  ><secondary
                    ><em>PTM Overview</em>: Tooltips, Highlighting, and Zooming </secondary
                  ></span
                >  
              </div>
            </small>
            <hr />
            <h6 id="about-c5"><strong>5 Shared PTM sites between modification types</strong></h6>
            <br>
            <div class="img-container thumbnail" style="width: 30%;">
              <video controls="controls" style="width: 100%;" name="Video Name" poster="{{ url_for('static', filename='resources/usage_guide_images/Shared_PTM_sites.png') }}">
                <source src="{{ url_for('static', filename='resources/usage_guide_movies/shared_sites_zooming.mov') }}"">
              </video>
              <span class="title"
                ><secondary
                  ><em>Shared PTM sites between modification types</em>: Tooltips, Highlighting, and Zooming
                  </secondary
                ></span
              >
            </div>            
            <br>
            <small>
              For each pair of modification types, this visualisation shows if 
              both types were found on common sites. If there are sites that were
              modified by both modifications within the data set,
              the number of sites is displayed with a color gradient, where 
              brighter cells indicate lower numbers of shared sites. 
              <br> 
              Note that the absolute number of shared sites is given and therefore not 
              normalized by the total number of sites for each modification type,
              meaning that more frequently occuring modifications will have
              more shared sites by chance.
              <br>
              Zooming in and out is possible by scrolling the mouse wheel. 
              <br>
              The visualisation can be panned by dragging the mouse. 
              <br> 
              The slider on the bottom left shows the range of the number of 
              shared sites that are displayed. Hovering over the slider will 
              highlight the corresponding cells in the visualisation.
            </small>
            <br>
            <hr />
            <h6 id="about-c6"><strong>6 Mass Shift</strong></h6>
            <br>
            <div class="img-container thumbnail" style="width: 30%;">
              <video controls="controls" width="600" height="600" name="Video Name" poster="{{ url_for('static', filename='resources/usage_guide_images/Mass_shift.png') }}">
                <source src="{{ url_for('static', filename='resources/usage_guide_movies/switch_sorting.mov') }}"">
              </video>
              <span class="title"
                ><secondary
                  ><em> Mass shift</em>: Tooltips, Highlighting, Zooming</secondary
                ></span
              >
            </div>     
            <br>
            <small>
              Mass shifts of each modification type are shown in the <em>Mass Shift</em>
              Scatter plot. Hovering over a point will display the mass shift  
              in Dalton. Depending on the input file, the mass shift is either 
              parsed directly from the file or retrieved from the UniMod database 
              (monoisotopic mass).
              Using the <i class="fa-duotone fa-arrow-up-arrow-down"></i> button 
              at the top right, the visualisation can be sorted by mass shift
              instead of Site Count.
              <br>
              In this sorted view, red bars are displayed to indicate mass shifts
              that are less than 0.02 Da apart to flag modifications that might 
              not be resolvable due to the mass accuracy of the mass spectrometer.
              Note that this is only a default value and the detectability
              is dependent on the mass accuracy of the mass spectrometer used, 
              as well as the precursor mass.
              <br>
              The mass shift scatter plot can be zoomed in and out by scrolling the mouse wheel. 
              <br>
              The visualisation can be panned by dragging the mouse. 
            </small>
            <hr />
            <h6 id="about-c7"><strong>7 Site Count</strong></h6>
            <br>
                <div class="img-container thumbnail" style="width: 30%;"">
                  <img
                    src="{{ url_for('static', filename='resources/usage_guide_images/Site_Count.png') }}"
                    alt="Example of the Site Count Histogram."
                  />
                  <span class="title"
                    ><secondary
                    >Example of the <em>Site Count Histogram</em>.</secondary
                    ></span
                  >
                </div>    
            <br>
            <small>
              The <em>Site Count</em> histogram shows the number of sites that 
              were modified by each modification type.
              Hovering over the bars will display the number of sites, 
              and additionally show you across how many proteins these sites are spreaded.
              <br>
              The histogram can be zoomed in and out by scrolling the mouse wheel. 
              <br>
              The visualisation can be panned by dragging the mouse.
            </small>
            <hr />
            <h6 id="about-c8"><strong>8 PTM Class Counts</strong></h6>
            <br>
            <div class="img-container thumbnail" style="width: 30%;"">
              <img
                src="{{ url_for('static', filename='resources/usage_guide_images/PTM_Class_counts.png') }}"
                alt="Example of the PTM Class Count Barchart."
              />
              <span class="title"
                ><secondary
                >Example of the <em>PTM Class Count Barchart</em>, showing the modification class tooltip.</secondary
                ></span
              >
            </div>    
            <br>
            <small>
              The <em>PTM Class Counts</em> barchart shows how many times
              each of the UniMod Classes have been assigned to a 
              modification + site pair.
              Note that UniMod classifications are depending on the 
              residue that is modified.
              Hovering over the bars will display the exact number of 
              modifications. 
            </small>
            <hr />
            <h6 id="about-c9"><strong>9 Overview Table</strong></h6>
            <small>
              The overview table displays all proteins present in the data set
              (one per row) after successful processing. In addition to the
              UniProt primary accession identifier and name matched from
              UniProt, the number of positions with at least one PTM and the
              number of different PTMs are displayed. The table can be sorted by
              all columns in ascending or descending order by clicking on them.
              The two input fields at the top allow filtering for PTM and/or
              protein name. For the former, only those proteins are displayed
              for which the corresponding PTM was detected. The table allows the
              selection of a single protein; this is indicated by red
              highlighting. After selection, the
              <em>Detail Visualizations</em> for the corresponding protein can
              be loaded by clicking on the
              <span
                style="
                  background-color: #ff6663;
                  padding: 0.5px;
                  margin: 1px;
                  font-size: small;
                  border-radius: 4px;
                  color: #fbfbfb;
                "
                >Display</span
              >
              button.
              <div class="img-container rounded thumbnail mt-4 mb-4">
                <img
                  src="{{ url_for('static', filename='resources/usage_guide_images/overview_table.png') }}"
                  alt="Display of the protein overview table with the PTM filter applied to oxidation, dioxidation and trioxidation. The protein P0ABB8 is selected."
                />
                <span class="title"
                  ><secondary
                    >Display of the protein overview table with the PTM filter
                    applied to <em>oxidation</em>, <em>dioxidation</em> and
                    <em>trioxidation</em>. The protein <em>P0ABB8</em> is
                    selected.</secondary
                  ></span
                >
              </div>
            </small>
            <hr />
            <h6 id="about-c10"><strong>10 Detail Visualizations</strong></h6>
            <small>
              PTMVision offers two distinct modes to display the PTM landscape
              of a single protein; The
              <em>Modifications View</em> (default, details are described in
              section 11) and the <em>Structure View</em> (details are described
              in section 12). Both are rendered in panel
              <i class="fa-duotone fa-circle-4"></i> and are compiled as soon as
              a single protein is selected for detailed reporting (see section
              nine). The currently selected protein (primary accession, name and
              source organism) is indicated in the header of panel
              <i class="fa-duotone fa-circle-4"></i> and can be clicked to
              retrieve UniProt comments for the protein. Furthermore, a range of
              user interactions is accessible via the buttons at the top right
              of panel <i class="fa-duotone fa-circle-4"></i>:
              <p>
                <i class="fa-duotone fa-camera"></i> Screenshot of the current
                view of the detail visualizations. The corresponding image file
                is automatically downloaded in <code>png</code> format.
              </p>
              <p>
                <i class="fa-duotone fa-magnifying-glass-arrows-rotate"></i>
                Reset the zoom level of all visualizations to default (zoomed
                out completely).
              </p>
              <p>
                <i class="fa-duotone fa-highlighter"></i> Highlight a specific
                PTM of interest. A pop-up window that supports search input
                prompts you to select a number of modifications to be
                highlighted. To remove a highlight, the interaction must be
                performed without selecting a modification. For the
                <em>Modifications View</em>, highlighting is applied in the same
                way as for the overview visualizations. For the
                <em>Structure View</em>, highlighting is applied to the protein
                structure model and those contacts, that comprise the respective
                PTMs.
              </p>
              <p>
                <i class="fa-duotone fa-repeat"></i> Switch between
                <em>Modifications View</em> and <em>Structure View</em>.
              </p>
            </small>
            <hr />
            <h6 id="about-c11"><strong>11 Modifications View</strong></h6>
            <small>
              <div class="img-container rounded thumbnail mt-4 mb-4">
                  <video controls="controls" style="width: 100%;" poster="{{ url_for('static', filename='resources/usage_guide_images/dashboard_modifications.png') }}" >
                    <source src="{{ url_for('static', filename='resources/usage_guide_movies/modification_view.mov') }}"">
                  </video>
                  <span class="title"
                    ><secondary
                      >Clip showing how to navigate the <em>Modification dashboard</em>.
                      </secondary
                    ></span
                  >
              </div>
              The <em>Modifications View</em> provides an overview over which modification types 
              are found in the previously selected protein and where they are located on the 
              primary sequence. The core visualisation is the 
              presence/absence matrix, where each row represents a modification type and each column
              a position in the protein sequence. A filled cell indicates that the modification type
              was found at the respective position. The cells are colored by the UniMod classification,
              the color legend is displayed on the top left side and can be used to filter for
              specific classes. 
              Filtering can be reset by clicking the "Select all" button, the current selection 
              can be inverted via the "Invert selection" button.
              <br>
              The histogram above the presence/absence plot shows the total number of modifications per
              position, the distribution of UniMod classes is encoded by colors. 
              Hovering over a bar will show the exact PTM class counts.
              This histogram can be used to navigate directly to a site of interest. 
              The histogram to the far right shows the total number of modification sites per modification type,
              the distribution of UniMod classes is encoded by colors as well. 
              Hovering over a bar will show the exact PTM class counts.
              This histogram can be used to navigate directly to a modification of interest. 

              <br>
              Tracks below the presence/absence plot show information that could be mapped via UniProt:
              <ul>
                <li>
                  <a
                    href="https://www.uniprot.org/help/disease_phenotypes_variants_section"
                    target="_blank"
                  >Experimental info: 
                  </a>
                  information on the disease(s), phenotype(s), and variants associated with the protein 
                </li>
                <li>
                  <a
                    href="https://www.uniprot.org/help/binding"
                    target="_blank"
                  >Sites: 
                  </a>
                  information on the interaction between protein residues and chemical entities (e.g. metals, cofactors)
                </li>
                <li>
                  <a
                    href="https://www.uniprot.org/help/family_and_domains_section"
                    target="_blank"
                  >Region: 
                  </a>
                  sequence similarities with other proteins and the domain(s) present in the protein
                </li>
                <li>
                  <a
                    href="https://www.uniprot.org/help/structure_section"
                    target="_blank"
                  >Secondary Structure: 
                  </a>
                  information on the secondary structure elements within the protein 
                </li>
                <li>
                  <a
                    href="https://www.uniprot.org/help/disease_phenotypes_variants_section"
                    target="_blank"
                  >Amino Acid Modifications: 
                  </a>
                  post-translational modifications (PTMs) and/or processing events
                </li>
                <li>
                  <a
                    href="https://www.uniprot.org/help/disease_phenotypes_variants_section"
                    target="_blank"
                  >Molecule Processing: 
                  </a>
                  information on the disease(s), phenotype(s), and variants associated with the protein 
                </li>
              </ul>                
              The heatmap to the right shows the distribution of modification types across the different
              amino acid residues. A colored cell indicates that the respective modification type was found
              on the respective amino acid. The color legend is displayed on the bottom, where a darker color
              indicates a higher number of sites with this modification-amino acid combination. 
              Note that the absolute, non-normalized numbers of sites are given.

            </small>
            <hr />
            <h6 id="about-c12"><strong>12 Structure View</strong></h6>
            <small>
              <div class="img-container rounded thumbnail mt-4 mb-4">
                <video controls="controls" style="width: 100%;" poster="{{ url_for('static', filename='resources/usage_guide_images/dashboard_structure.png') }}">
                  <source src="{{ url_for('static', filename='resources/usage_guide_movies/structure_view.mov') }}">
                </video>
                <span class="title"
                  ><secondary
                    >Clip showing how to navigate the <em>Structure dashboard</em>.
                    </secondary
                  ></span
                >  
              </div>
              Clicking the <i class="fa-duotone fa-repeat"></i> switch button on the panel menu will open 
              the structure dashboard. 
              <br>
              The left side shows the 3D structure in 2D via a contact map. The residues are plotted against 
              each other, and a cell is colored if the respective residues are in close contact, 
              meaning that their C-beta (Glycine: C-alpha) atoms are within x Angström. 
              The cells are colored according to the modification state of the respective residue pair:
              <ul>

              <li>Light grey cells represent pairs of amino acids that are in contact, and both residues are unmodified </li>
              <li>Dark grey cells represent pairs of amino acids that are in contact, and one or both residues are modified, 
              but no modification type is shared between the residues </li>
              <li>Red cells represent pairs of amino acids that are in contact, both residues are modified, and they 
              share at least one modification type (e.g. both were found to be phosphorylated) </li>
              <li>If a modification was highlighted via the <i class="fa-duotone fa-highlighter"></i> button in the
              menu, residues that are in contact and where one or both carry the selected modification are 
              colored in blue. </li>
              </ul>
              Via the color legend on the top, the classes can be filtered. The filtering can be reset 
              via the "Select all" button, the current selection can be inverted via the "Invert selection" button.

              The histogram on top and the tracks at the bottom are the same as in the <em>Modifications View</em>:
              The first shows the total number of modifications per position, the second shows 
              the annotations mapped via UniProt (see section 11).

              <br>
              Hovering over a cell in the contact map shows additional info on the two residues (PTM class counts,
              UniProt Annotations). Clicking on a cell will highlight the respective residues in the structure
              model on the right side, and show their distance in Angström. 

              Highlighting a modification via the <i class="fa-duotone fa-highlighter"></i> button will color the
              residues in the structure model that carry the selected modification in blue.
              The structure model can be rotated by clicking and dragging the mouse, zoomed in and out by scrolling
              the mouse wheel, and panned by dragging the mouse.
              
              <br>
              Clicking on a cell will show a track at the bottom of the structure model that displays the
              modifications of the respective residues, plotted according to their mass shift in Dalton. 

            </small>
          </div>
        </div>

        <div class="frame">
          <div class="heading">FAQ</div>
          <div class="content">
            We currently have no content for this section. If you encounter any
            problems or have any questions, do not hesitate to
            <a
              href="https://github.com/Integrative-Transcriptomics/PTMVision/issues"
              target="_blank"
              >open an issue in our GitHub repository</a
            >.
          </div>
        </div>

        <div class="frame">
          <div class="heading">Legal Information</div>
          <div class="content">
            <dl>
              <small>
                <dt class="infotext-header">Cookie usage</dt>
                <dd class="infotext-small">
                  We use a strictly necessary (http-only non-tracking) session
                  cookie so that you can access your data. (1) Your data will
                  not be accessible to other people. (2) No personalized data is
                  collected or shared with third parties and you will not be
                  tracked by this cookie. (3) The cookie will expire after you
                  close your tab.
                </dd>
                <dt class="infotext-header">Content liability</dt>
                <dd class="infotext-small">
                  The contents of our pages were created with the utmost care.
                  However, we cannot guarantee the accuracy, completeness and
                  actuality of the content. As a service provider, we are
                  responsible for our own content on these pages under the
                  general laws according to § 7 TMG. According to §§ 8 to 10
                  TMG, however, we are not obligated as a service provider to
                  monitor transmitted or stored third-party information or to
                  investigate circumstances that indicate illegal activity.
                  Obligations to remove or block the use of information under
                  the general laws remain unaffected. However, liability in this
                  regard is only possible from the point in time at which a
                  concrete infringement of the law becomes known. If we become
                  aware of such infringements, we will remove this content
                  immediately.
                </dd>
                <dt class="infotext-header">Link liability</dt>
                <dd class="infotext-small">
                  This page may contain links to external websites of third
                  parties, on whose contents we have no influence. We cannot
                  take any liability for these external contents. The respective
                  provider or operator of the pages is always responsible for
                  the content of the linked pages. The linked pages were checked
                  for possible legal violations at the time of linking. Illegal
                  contents were not recognizable at the time of linking.
                  However, a permanent control of the contents of the linked
                  pages is not reasonable without concrete evidence of a
                  violation of the law. If we become aware of any infringements,
                  we will remove such links immediately.
                </dd>
                <dt class="infotext-header">Data privacy</dt>
                <dd class="infotext-small">
                  The use of our website is possible without providing personal
                  data and no processed data is stored. If personal data (such
                  as name, address or e-mail addresses) is collected, this is on
                  a voluntary basis. This data will not be passed on to third
                  parties without your express consent. We point out that data
                  transmission over the Internet (e.g. communication by e-mail)
                  may yield security gaps. Complete protection of data against
                  access by third parties is not possible. The use of contact
                  data published within the framework of the impressum by third
                  parties for the purpose of sending advertising and information
                  material is hereby expressly prohibited. The operators of the
                  pages expressly reserve the right to take legal action in the
                  event of the sending of advertising information, such as spam
                  e-mails.
                </dd>
                <dt class="infotext-header">Copyright</dt>
                <dd class="infotext-small">
                  This program is free software: you can redistribute it and/or
                  modify it under the terms of the GNU General Public License as
                  published by the Free Software Foundation, either version 3 of
                  the License, or (at your option) any later version. This
                  program is distributed in the hope that it will be useful, but
                  WITHOUT ANY WARRANTY; without even the implied warranty of
                  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
                  GNU General Public License for more details. (<a
                    href="https://www.gnu.org/licenses/gpl-3.0.en.html"
                    target="_blank"
                    >www.gnu.org/licenses/gpl-3.0.en</a
                  >)
                </dd>
                <dt class="infotext-header">Code availability</dt>
                <dd class="infotext-small">
                  The source code of PTMVision is available at
                  <a
                    href="https://github.com/Integrative-Transcriptomics/PTMVision"
                    target="_blank"
                    >github.com/Integrative-Transcriptomics/PTMVision</a
                  >.
                </dd>
                <dt class="infotext-header">Contact</dt>
                <dd class="infotext-small">
                  <div>
                    <address>
                      Caroline Jachmann
                      <br />
                      Computational Omics and Systems Biology Group, Department
                      of Biomolecular Medicine, Ghent University
                      <br />
                      VIB Technologiepark 75<br />
                      9052 Gent-Zwijnaarde, Belgium<br />
                      <a href="mailto:caroline.jachmann@ugent.be"
                        >caroline.jachmann@ugent.be</a
                      >
                    </address>
                    <address>
                      Simon Hackl
                      <br />
                      Integrative Transcriptomics Group, Institute fpr
                      Bioinformatics and Medical Informatics, University
                      Tübingen
                      <br />
                      Sand 14<br />
                      72076 Tübingen, Germany<br />
                      <a href="mailto:simon.hackl@uni-tuebingen.de"
                        >simon.hackl@uni-tuebingen.de</a
                      >
                    </address>
                  </div>
                </dd>
              </small>
            </dl>
          </div>
        </div>
      </div>
    </div>
  </body>
</html>
