<!DOCTYPE html>
<!--
PTMVision WEB APPLICATION
AUTHOR: Caroline Jachmann, Simon Hackl
CONTACT: caroline.jachmann@uni-tuebingen.de, simon.hackl@uni-tuebingen.de
VERSION: 1.0

LICENSE
This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.
This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.
You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.
-->
<html lang="en">
  <head>
    <title>PTMVision</title>
    <link
      rel="icon"
      type="image/x-icon"
      href="{{ url_for('static', filename='resources/icon.png') }}"
    />
    <!--
      Provide meta information about the application.
      -->
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <meta
      name="description"
      content="PTMVision: Interactive Visualization of Post Translational Modifications | Understanding the intricate landscape of PTMs is crucial for unraveling the complex mechanisms underlying cellular processes and diseases. With PTMVision, researchers can effortlessly analyze and explore PTM data derived from various sources, including search engines such as ionbot, MaxQuant, MSFragger, Sage, MaxQuant, or from a plain CSV table format processed from mass spectrometry (MS) data."
    />
    <meta name="author" content="Caroline Jachmann, Simon Hackl" />
    <!--
      Definition of stylesheets to use in the application.
      -->
    <link
      rel="stylesheet"
      href="{{url_for('static', filename='css/metro-5.0.0.rc9.min.css')}}"
    />
    <link
      rel="stylesheet"
      type="text/css"
      href="{{url_for('static', filename='css/tabulator-6.1.0.min.css')}}"
    />
    <link
      rel="stylesheet"
      type="text/css"
      href="{{url_for('static', filename='css/tabulator-bootstrap4-6.1.0.min.css')}}"
    />
    <link
      rel="stylesheet"
      type="text/css"
      href="{{url_for('static', filename='css/ptmvision.index-1.1.css')}}"
    />
    <!--
      Definition of java script files to use in the application.
      -->
    <script
      type="text/javascript"
      src="{{url_for('static', filename='js/jquery-3.7.1.min.js')}}"
    ></script>
    <script
      type="text/javascript"
      src="{{url_for('static', filename='js/metro-5.0.0.rc9.min.js')}}"
    ></script>
    <script
      type="text/javascript"
      src="{{url_for('static', filename='js/resizesensor-0.0.6.min.js')}}"
    ></script>
    <script
      type="text/javascript"
      src="{{url_for('static', filename='js/echarts-5.5.0.min.js')}}"
    ></script>
    <script
      type="text/javascript"
      src="{{url_for('static', filename='js/3Dmol-2.1.0.min.js')}}"
    ></script>
    <script
      type="text/javascript"
      src="{{url_for('static', filename='js/sweetalert2-11.10.8.min.js')}}"
    ></script>
    <script
      type="text/javascript"
      src="{{url_for('static', filename='js/axios-1.1.2.min.js')}}"
    ></script>
    <script
      type="text/javascript"
      src="{{url_for('static', filename='js/lodash-4.17.21.min.js')}}"
    ></script>
    <script
      type="text/javascript"
      src="{{url_for('static', filename='js/tabulator-6.1.0.min.js')}}"
    ></script>
    <script
      type="text/javascript"
      src="{{url_for('static', filename='js/pako-2.1.0.min.js')}}"
    ></script>
    <script
      type="text/javascript"
      src="{{url_for('static', filename='js/ptmvision.index-1.1.js')}}"
    ></script>
    <script
      type="text/javascript"
      src="https://kit.fontawesome.com/df04f22e3d.js"
      crossorigin="anonymous"
    ></script>
  </head>
  <body onLoad="init();">
    <!-- App menu -->
    <header
      id="menu"
      class="app-bar app-bar-expand"
      data-role="appbar"
      data-expand="true"
      data-role-appbar="true"
    >
      <div id="menu-grid" class="grid w-100">
        <div class="row">
          <div class="cell text-left ml-4 w-100">
            <div class="logo no-hover">
              <img
                src="{{ url_for('static', filename='resources/logo.png') }}"
                alt="PTMVision Logo"
              />
            </div>
          </div>
          <div class="cell text-right">
            <button
              onclick="downloadSessionData()"
              class="button-no-decoration menu-button mr-8"
              data-role="hint"
              data-hint-text="Download (binary) session data for later re-upload."
              data-hint-position="left"
            >
              <i class="fa-duotone fa-file-arrow-down fa-2xl"></i>
            </button>
            <button
              onclick="redirectAbout()"
              class="button-no-decoration menu-button mr-8"
              data-role="hint"
              data-hint-text="Redirect to the PTMVision help and about page."
              data-hint-position="left"
            >
              About
            </button>
          </div>
        </div>
      </div>
    </header>

    <div id="main">
      <div id="panel">
        <div id="vline"></div>

        <div id="panel-info" class="panel-container">
          <h5>
            PTMVision: Interactive Visualization of Post Translational
            Modifications
          </h5>
          <p class="text-just text-light">
            Understanding the intricate landscape of post-translational
            modifications (PTMs) is crucial for unraveling the complex
            mechanisms underlying cellular processes and diseases. With
            PTMVision, researchers can effortlessly analyze and explore PTM data
            derived from various proteomics search engines, including ionbot,
            MSFragger, Sage, or from a plain CSV table format processed manually
            from mass spectrometry (MS) data.
            <b
              >For a detailed explanation of the use of PTMVision and
              interpretation of the visualizations offered, please refer to the
              <a href="#" onclick="redirectAbout()">About</a>
              page of PTMVision.</b
            >
          </p>
          <dl>
            <dt>
              <h4>
                <i class="fa-solid fa-person-chalkboard"></i> Short Usage
                Instruction
              </h4>
            </dt>
            <small>
              <dt class="infotext-header">
                <i class="fa-duotone fa-circle-1"></i> Upload data
              </dt>
              <dd class="infotext-small">
                <ul style="list-style-position: inside; padding-left: 0">
                  <li>
                    Upload your data or re-upload a previously downloaded
                    PTMVision session file at panel
                    <i class="fa-duotone fa-circle-1 fa-sm"></i> by entering all
                    required inputs and clicking on
                    <span
                      style="
                        background-color: #ff6663;
                        padding: 2px;
                        margin: 1px;
                        font-size: small;
                        border-radius: 4px;
                        color: #fbfbfb;
                      "
                      >Submit</span
                    >. The
                    <i class="fa-duotone fa-solid fa-circle-info"></i> icons on
                    the single input fields can be hovered for more information.
                  </li>
                  <li>
                    Once the processing is complete, your current session data
                    can be downloaded by clicking on
                    <i class="fa-duotone fa-file-arrow-down"></i> in the header
                    menu on the right.
                  </li>
                  <li>
                    To explore the features of PTMVision,
                    <b>example sessions</b> can be started below:
                  </li>
                </ul>
              </dd>
              <!-- Add example sessions here, if new ones are available. -->
              <table class="table striped w-75 ml-12">
                <thead>
                  <tr>
                    <th>Search Engine</th>
                    <th>Version</th>
                    <th>Search Type</th>
                    <th>Source Data</th>
                    <th>Note</th>
                    <th></th>
                    <th></th>
                  </tr>
                </thead>
                <tbody>
                  <tr>
                    <td>Ionbot</td>
                    <td>v0.11.0</td>
                    <td>Open</td>
                    <td>
                      PRIDE project
                      <a
                        href="https://www.ebi.ac.uk/pride/archive/projects/PXD000498"
                        >PXD000498</a
                      >, LC-MS/MS run of an unenriched
                      <i>Escherichia coli</i> sample.
                    </td>
                    <td>
                      Unimod classes
                      <i>Artefact</i> and <i>Chemical derivative</i> excluded
                      from analysis.
                    </td>
                    <td>
                      <button class="example-button">Start Session</button>
                    </td>
                    <td>
                      <button class="example-button">
                        Download Input Data
                      </button>
                    </td>
                  </tr>
                  <tr>
                    <td>MSFragger</td>
                    <td>v4.0</td>
                    <td>Open</td>
                    <td>
                      PRIDE project
                      <a
                        href="https://www.ebi.ac.uk/pride/archive/projects/PXD007740"
                        >PXD007740</a
                      >, LC-MS/MS run of a human colorectal cancer sample
                      enriched for phosphorylated peptides using
                      Ti<sup>4+</sup>-IMAC.
                    </td>
                    <td>
                      Source file filtered for modified peptides, Unimod classes
                      <i>Artefact</i> and <i>Chemical derivative</i> excluded
                      from analysis.
                    </td>
                    <td>
                      <button class="example-button">Start Session</button>
                    </td>
                    <td>
                      <button class="example-button">
                        Download Input Data
                      </button>
                    </td>
                  </tr>
                  <tr>
                    <td>Sage</td>
                    <td>v0.13.1</td>
                    <td>Closed</td>
                    <td>
                      PRIDE project
                      <a
                        href="https://www.ebi.ac.uk/pride/archive/projects/PXD000498"
                        >PXD000498</a
                      >, LC-MS/MS run
                        of an unenriched <i>Escherichia coli</i> sample</a
                      >.
                    </td>
                    <td></td>
                    <td>
                      <button class="example-button">Start Session</button>
                    </td>
                    <td>
                      <button class="example-button">
                        Download Input Data
                      </button>
                    </td>
                  </tr>
                  <tr>
                    <td>MaxQuant</td>
                    <td>v2.4.13.0</td>
                    <td>Closed</td>
                    <td>
                      PRIDE project
                      <a
                        href="https://www.ebi.ac.uk/pride/archive/projects/PXD025088"
                        >PXD025088</a
                      >, LC-MS/MS run
                        of an unenriched <i>Escherichia coli</i> sample</a
                      >.
                    </td>
                    <td>
                      Source file filtered for modified peptides.
                    </td>
                    <td>
                      <button class="example-button">Start Session</button>
                    </td>
                    <td>
                      <button class="example-button">
                        Download Input Data
                      </button>
                    </td>
                  </tr>
                  <tr>
                    <td>Plain CSV</td>
                    <td></td>
                    <td>Open</td>
                    <td>
                      <a
                        href="http://biovis.net/2022/biovisChallenges_vis/"
                        >IEEE VIS 2022 Bio+MedVis challenge dataset</a
                      > comprising  modifications of three human proteins and their mouse orthologs identified with Ionbot.
                    </td>
                    <td>
                      Unimod classes
                      <i>Artefact</i> and <i>Chemical derivative</i> excluded
                      from analysis.
                    </td>
                    <td>
                      <button class="example-button">Start Session</button>
                    </td>
                    <td>
                      <button class="example-button">
                        Download Input Data
                      </button>
                    </td>
                  </tr>
                </tbody>
              </table>
              <br />
              <dt class="infotext-header">
                <i class="fa-duotone fa-circle-2"></i> Overview
                post-translational modifications
              </dt>
              <dd class="infotext-small">
                <ul style="list-style-position: inside; padding-left: 0">
                  <li>
                    After the data transfer and successful processing, you will
                    be provided with a dashboard with an overview of all PTMs in
                    your data in panel
                    <i class="fa-duotone fa-circle-2 fa-sm"></i>.
                  </li>
                  <li>
                    This allows you to obtain a quantitative and qualitative
                    assessment of the PTMs in your data, independent of
                    individual proteins.
                  </li>
                </ul>
              </dd>
              <br />
              <dt class="infotext-header">
                <i class="fa-duotone fa-circle-3"></i> Overview proteins and
                select proteins of interest
              </dt>
              <dd class="infotext-small">
                <ul style="list-style-position: inside; padding-left: 0">
                  <li>
                    A tabular overview of all proteins available from your data
                    is given in panel
                    <i class="fa-duotone fa-circle-3 fa-sm"></i>, listing one
                    protein per row with basic information about the protein and
                    the associated PTMs.
                  </li>
                  <li>
                    The rows of the table can be filtered according to the name
                    of the protein, the UniProt identifier and the presence of
                    modifications to proteins based on their names. All filters
                    are applied in an AND logic.
                  </li>
                  <li>
                    Select a single row from the table by clicking on it (it
                    will be highlighted in red) and click
                    <span
                      style="
                        background-color: #ff6663;
                        padding: 2px;
                        margin: 1px;
                        font-size: small;
                        border-radius: 4px;
                        color: #fbfbfb;
                      "
                      >Display</span
                    >
                    to obtain a detailed view of the PTMs in the selected
                    protein in panel
                    <i class="fa-duotone fa-circle-4 fa-sm"></i>.
                  </li>
                </ul>
              </dd>
              <br />
              <dt class="infotext-header">
                <i class="fa-duotone fa-circle-4"></i> Explore the details of
                selected proteins
              </dt>
              <dd class="infotext-small">
                Detail information about the occurrence of PTMs for the selected
                protein is displayed in panel
                <i class="fa-duotone fa-circle-4 fa-sm"></i>. You can switch
                between two display modes via the top-right button on the panel:
                <br />
                <span>&#8226;</span> The <b>Modification View</b> focuses on the
                occurrence of PTMs along the selected protein primary sequence
                as well as per amino acid.
                <br />
                <span>&#8226;</span> The <b>Structure View</b> focuses on the
                occurrence of PTMs in the context of residues that are in close
                spatial contact.
              </dd>
            </small>
            <br />
            <dt>
              <h4>
                <i class="fa-solid fa-triangle-exclamation"></i> Disclaimer
              </h4>
            </dt>
            <small>
              <dd class="infotext-small">
                <span>&#8226;</span> We recommend using PTMVision in Chrome or
                Firefox for optimal performance. Safari has issues when
                visualising larger datasets.
                <br />
                <span>&#8226;</span> Note that PTMVision serves solely to
                visualize the content provided in the input and does not conduct
                any further quality control. The localization of modifications
                still remains a challenging task and may yield imprecise
                results.
                <br />
                <span>&#8226;</span> PTMVision will be continuously updated and
                improved, including the addition of filters for localisation
                scores as soon as they will be reported by search engines.
              </dd>
            </small>
          </dl>
        </div>

        <div id="panel-inputs" class="panel-container">
          <h6 class="secondary">
            <i class="fa-duotone fa-circle-1"></i>
            <strong
              >Upload post-translational modification data from open or closed
              search <u>or</u> Restart a previous PTMVision session</strong
            >
          </h6>
          <h7 class="secondary"
            ><b style="color: #dc5754">Required input: </b>Specify the
            <b>pipeline/proteomics search engine</b> with which your data was
            generated. This defines the data format to be processed.</h7
          >
          <select id="data-type-form" class="mb-2" data-role="select">
            <optgroup label="Pipeline">
              <option value="ionbot">Ionbot</option>
              <option value="msfragger">MSFragger</option>
              <option value="sage">Sage (closed)</option>
              <option value="msms">MaxQuant (closed)</option>
              <option value="csv">Plain CSV</option>
              <!-- <option value="infer">Infer (Experimental)</option> -->
            </optgroup>
          </select>
          <h7 class="secondary">
            <b style="color: #dc5754">Required input: </b>Specify the
            <b>PTM data</b> file. This is either the output of the chosen
            proteomics search engine or a manually compiled file (Plain
            CSV).</h7
          >
          <input
            id="data-input-form"
            class="mb-2"
            type="file"
            data-role="file"
            data-button-title='<i class="fa-duotone fa-folder-open"></i>'
          />
          <h7 class="secondary"
            ><b style="color: #666666">Optional selection: </b> Specify (Unimod)
            modification classes to exclude from processing. Any PTM in your
            data of one of the selected classes will be ignored.</h7
          >
          <select
            id="data-excludecls-form"
            class="mb-2"
            data-role="select"
            multiple
          >
            <optgroup label="Other">
              <option value="Chemical derivative">Chemical derivative</option>
              <option value="Artefact">Artefact</option>
              <option value="Synth. pep. protect. gp.">
                Synth. pep. protect. gp.
              </option>
              <option value="Isotopic label">Isotopic label</option>
              <option value="Non-standard residue">Non-standard residue</option>
              <option value="Multiple">Multiple</option>
              <option value="Other">Other</option>
              <option value="AA substitution">AA substitution</option>
              <option value="Deprecated unimod entry">
                Deprecated unimod entry
              </option>
            </optgroup>
            <optgroup label="Translational">
              <option value="Post-translational">Post-translational</option>
              <option value="Co-translational">Co-translational</option>
              <option value="Pre-translational">Pre-translational</option>
            </optgroup>
            <optgroup label="Glycosylation">
              <option value="N-linked glycosylation">
                N-linked glycosylation
              </option>
              <option value="O-linked glycosylation">
                O-linked glycosylation
              </option>
              <option value="Other glycosylation">Other glycosylation</option>
            </optgroup>
            <optgroup label="Cross-Link">
              <option value="Cross-link">Cross-link</option>
              <option value="CID cleavable cross-link">
                CID cleavable cross-link
              </option>
              <option value="Photo cleavable cross-link">
                Photo cleavable cross-link
              </option>
              <option value="Other cleavable cross-link">
                Other cleavable cross-link
              </option>
            </optgroup>
          </select>
          <h7 class="secondary">
            <b style="color: #666666">Optional selection: </b> Specify the mass
            shift tolerance. This is only relevant for closed search data and
            will merge PTMs within the specified tolerance.
          </h7>
          <br />
          <input
            id="data-tolerance-form"
            class="mt-1 mb-1 w-25 input-small"
            type="number"
            value="0.001"
            min="0"
            max="10"
            step="0.001"
          />
          <br />
          <button class="proceed-button mt-2" onclick="startSession()">
            Submit
          </button>
          <h5>or</h5>
          <h7 class="secondary"
            ><b style="color: #dc5754">Required input: </b>Specify a previously
            downloaded PTMVision session <code>.zlib</code> file.</h7
          >
          <input
            id="session-input-form"
            class="mb-1"
            type="file"
            data-role="file"
            accept=".zlib"
            data-button-title='<i class="fa-duotone fa-folder-open"></i>'
          />
          <button class="proceed-button mt-2" onclick="startExistingSession()">
            Submit
          </button>
        </div>

        <div id="panel-overview" class="panel-container">
          <h6 class="secondary">
            <i class="fa-duotone fa-circle-2"></i>
            <strong>Overview PTMs</strong>
          </h6>
          <div id="panel-overview-chart"></div>
          <div id="panel-overview-controls">
            <button
              id="panel-overview-control-download"
              class="button-no-decoration"
              data-role="hint"
              data-hint-text="Download image."
              data-hint-position="left"
              onclick="overviewChartDownloadImage()"
            >
              <i class="fa-duotone fa-camera"></i>
            </button>
            <button
              id="panel-overview-control-restore"
              class="button-no-decoration"
              data-role="hint"
              data-hint-text="Reset zoom."
              data-hint-position="left"
              onclick="overviewChartRestoreZoom()"
            >
              <i class="fa-duotone fa-magnifying-glass-arrows-rotate"></i>
            </button>
            <button
              id="panel-overview-control-highlight"
              class="button-no-decoration"
              data-role="hint"
              data-hint-text="Highlight PTM."
              data-hint-position="left"
              onclick="overviewChartHighlight()"
            >
              <i class="fa-duotone fa-highlighter"></i>
            </button>
            <button
              id="panel-overview-control-sort"
              class="button-no-decoration"
              data-role="hint"
              data-hint-text="Sort by count/mass shift."
              data-hint-position="left"
              onclick="overviewChartSort()"
            >
              <i class="fa-duotone fa-arrow-up-arrow-down"></i>
            </button>
          </div>
        </div>

        <div id="panel-table" class="panel-container">
          <div class="grid">
            <div class="row">
              <div class="cell-11">
                <h6 class="secondary">
                  <i class="fa-duotone fa-circle-3"></i>
                  <strong id="panel-table-title"
                    >Select single protein of interest</strong
                  >
                </h6>
              </div>
            </div>
            <div class="row">
              <div class="cell-3">
                <h7 class="secondary float-left"
                  >Filter for modifications by UniMod name:&nbsp;</h7
                >
              </div>
              <div class="cell-7">
                <select
                  id="panel-table-filter-modification"
                  class="input-small"
                  data-role="select"
                  data-on-change="overviewTableSetFilters"
                  multiple
                ></select>
              </div>
            </div>
            <div class="row mt-2">
              <div class="cell-3">
                <h7 class="secondary float-left"
                  >Filter for protein names/UniProt identifier:&nbsp;</h7
                >
              </div>
              <div class="cell-7">
                <select
                  id="panel-table-filter-id"
                  class="input-small"
                  data-role="select"
                  data-on-change="overviewTableSetFilters"
                  multiple
                ></select>
              </div>
            </div>
          </div>
          <div id="panel-table-tabulator"></div>
          <button
            id="dashboard-display-button"
            class="proceed-button"
            onclick="dashboardChartInitialize( )"
            disabled
          >
            Display
          </button>
        </div>

        <div id="panel-dashboard" class="panel-container">
          <h6 class="secondary">
            <i class="fa-duotone fa-circle-4"></i>
            <strong id="panel-dashboard-title">Explore detail</strong>
            <u id="panel-dashboard-selection"></u>
          </h6>
          <div id="panel-dashboard-chart"></div>
          <div id="panel-dashboard-structure">
            <div id="panel-dashboard-structure-meta"></div>
          </div>
          <div id="panel-dashboard-controls">
            <button
              id="panel-dashboard-control-download"
              class="button-no-decoration"
              data-role="hint"
              data-hint-text="Download image (no structure)."
              data-hint-position="left"
              onclick="dashboardChartDownloadImage()"
            >
              <i class="fa-duotone fa-camera"></i>
            </button>
            <button
              id="panel-dashboard-control-restore"
              class="button-no-decoration"
              data-role="hint"
              data-hint-text="Reset zoom."
              data-hint-position="left"
              onclick="dashboardChartRestoreZoom()"
            >
              <i class="fa-duotone fa-magnifying-glass-arrows-rotate"></i>
            </button>
            <button
              id="panel-dashboard-control-highlight"
              class="button-no-decoration"
              data-role="hint"
              data-hint-text="Highlight PTM."
              data-hint-position="left"
              onclick="dashboardChartHighlight()"
            >
              <i class="fa-duotone fa-highlighter"></i>
            </button>
            <button
              id="panel-dashboard-control-switch"
              class="button-no-decoration"
              data-role="hint"
              data-hint-text="Switch between modifications/structure view."
              data-hint-position="left"
              onclick="dashboardChartSwitch()"
            >
              <i class="fa-duotone fa-repeat"></i>
            </button>
          </div>
        </div>
      </div>
    </div>
  </body>
</html>
